Flaky test: ProcessEventInspectionsTest.testFindLastProcessInstance

**Description**

It seems to be a timing issue. This process contains a timer and the test increases the time of the engine: `Utilities.increaseTime(engine, Duration.ofDays(1));`. Possibly the engine has not yet triggered the timer and started processing before we checked if it's idle. This means we consider the engine to be idle, even though it still has stuff to do (triggering the timer).

```
java.lang.AssertionError: 

Expecting Optional to contain a value but it was empty.
	at io.camunda.zeebe.process.test.qa.inspections.ProcessEventInspectionsTest.testFindLastProcessInstance(ProcessEventInspectionsTest.java:61)
```

Occurs with both extensions.

--- a/.github/workflows/build-test.yml
+++ b/.github/workflows/build-test.yml
@@ -38,7 +38,7 @@ jobs:
       fail-fast: false
       max-parallel: 1
       matrix:
-        os: [ macos-latest, windows-latest, ubuntu-latest ]
+        os: [ ubuntu-latest, macos-latest, windows-latest ]
 
     steps:
       - name: Checkout

--- a/engine-agent/pom.xml
+++ b/engine-agent/pom.xml
@@ -34,6 +34,15 @@
       <artifactId>slf4j-api</artifactId>
     </dependency>
 
+    <!--  Usually we don't want to provide an slf4j implementation in this project. This module
+      is the only exception to this rule. This is because we need to provide the engine with an
+      implementation, and since this runs in a docker container the user won't be providing it. -->
+    <dependency>
+      <groupId>org.slf4j</groupId>
+      <artifactId>slf4j-simple</artifactId>
+      <scope>runtime</scope>
+    </dependency>
+
     <dependency>
       <groupId>io.grpc</groupId>
       <artifactId>grpc-stub</artifactId>
@@ -79,6 +88,15 @@
             </configuration>
           </execution>
         </executions>
+      </plugin>
+
+      <plugin>
+        <groupId>org.apache.maven.plugins</groupId>
+        <artifactId>maven-dependency-plugin</artifactId>
+        <configuration>
+          <failOnWarning>true</failOnWarning>
+          <ignoreUnusedRuntime>true</ignoreUnusedRuntime>
+        </configuration>
       </plugin>
     </plugins>
   </build>

--- a/engine-agent/src/main/java/io/camunda/zeebe/process/test/engine/agent/RecordStreamSourceWrapper.java
+++ b/engine-agent/src/main/java/io/camunda/zeebe/process/test/engine/agent/RecordStreamSourceWrapper.java
@@ -2,12 +2,15 @@ package io.camunda.zeebe.process.test.engine.agent;
 
 import io.camunda.zeebe.process.test.api.RecordStreamSource;
 import java.util.ArrayList;
+import java.util.Collections;
 import java.util.List;
+import java.util.stream.StreamSupport;
 
 public class RecordStreamSourceWrapper {
 
   private final List<String> mappedRecords = new ArrayList<>();
   private final RecordStreamSource recordStreamSource;
+  private volatile long lastEventPosition = -1L;
 
   public RecordStreamSourceWrapper(final RecordStreamSource recordStreamSource) {
     this.recordStreamSource = recordStreamSource;
@@ -15,9 +18,15 @@ public class RecordStreamSourceWrapper {
 
   public List<String> getMappedRecords() {
     synchronized (mappedRecords) {
-      mappedRecords.clear();
-      recordStreamSource.records().forEach(record -> mappedRecords.add(record.toJson()));
+      StreamSupport.stream(recordStreamSource.records().spliterator(), false)
+          .filter(record -> record.getPosition() > lastEventPosition)
+          .forEach(
+              record -> {
+                mappedRecords.add(record.toJson());
+                lastEventPosition = record.getPosition();
+              });
     }
-    return new ArrayList<>(mappedRecords);
+
+    return Collections.unmodifiableList(mappedRecords);
   }
 }

--- a/qa/src/test/java/io/camunda/zeebe/process/test/qa/util/Utilities.java
+++ b/qa/src/test/java/io/camunda/zeebe/process/test/qa/util/Utilities.java
@@ -21,58 +21,6 @@ import java.util.stream.Collectors;
 
 public class Utilities {
 
-  public static final class ProcessPackLoopingServiceTask {
-    public static final String RESOURCE_NAME = "looping-servicetask.bpmn";
-    public static final String PROCESS_ID = "looping-servicetask";
-
-    public static final String ELEMENT_ID = "servicetask"; // id of the service task
-    public static final String JOB_TYPE = "test"; // job type of service task
-    public static final String TOTAL_LOOPS =
-        "totalLoops"; // variable name to indicate number of loops
-    public static final String GATEWAY_ELEMENT_ID = "Gateway_0fhwf5d";
-    public static final String START_EVENT_ID = "startevent";
-    public static final String END_EVENT_ID = "endevent";
-  }
-
-  public static final class ProcessPackMultipleTasks {
-    public static final String RESOURCE_NAME = "multiple-tasks.bpmn";
-    public static final String PROCESS_ID = "multiple-tasks";
-    public static final String ELEMENT_ID_1 = "servicetask1";
-    public static final String ELEMENT_ID_2 = "servicetask2";
-    public static final String ELEMENT_ID_3 = "servicetask3";
-  }
-
-  public static final class ProcessPackMessageEvent {
-    public static final String RESOURCE_NAME = "message-event.bpmn";
-    public static final String PROCESS_ID = "message-event";
-    public static final String MESSAGE_NAME = "message";
-    public static final String CORRELATION_KEY_VARIABLE = "correlationKey";
-  }
-
-  public static final class ProcessPackMessageStartEvent {
-    public static final String RESOURCE_NAME = "message-start-event.bpmn";
-    public static final String MESSAGE_NAME = "start-message";
-    public static final String CORRELATION_KEY = "";
-  }
-
-  public static final class ProcessPackTimerStartEvent {
-    public static final String RESOURCE_NAME = "timer-start-event-daily.bpmn";
-    public static final String TIMER_ID = "timer";
-  }
-
-  public static final class ProcessPackCallActivity {
-    public static final String RESOURCE_NAME = "call-activity.bpmn";
-    public static final String PROCESS_ID = "call-activity";
-    public static final String CALL_ACTIVITY_ID = "callactivity";
-    public static final String CALLED_RESOURCE_NAME = ProcessPackStartEndEvent.RESOURCE_NAME;
-    public static final String CALLED_PROCESS_ID = ProcessPackStartEndEvent.PROCESS_ID;
-  }
-
-  public static final class ProcessPackStartEndEvent {
-    public static final String RESOURCE_NAME = "start-end.bpmn";
-    public static final String PROCESS_ID = "start-end";
-  }
-
   public static DeploymentEvent deployProcess(final ZeebeClient client, final String process) {
     return deployProcesses(client, process);
   }
@@ -162,12 +110,13 @@ public class Utilities {
 
   public static void increaseTime(final InMemoryEngine engine, final Duration duration)
       throws InterruptedException {
-    engine.increaseTime(duration);
     try {
-      waitForBusyState(engine, Duration.ofSeconds(5));
       waitForIdleState(engine, Duration.ofSeconds(1));
-    } catch (TimeoutException e) {
-      // Do nothing. We've waited up to 250 ms for processing to start, if it didn't start in this
+      engine.increaseTime(duration);
+      waitForBusyState(engine, Duration.ofSeconds(1));
+      waitForIdleState(engine, Duration.ofSeconds(1));
+    } catch (final TimeoutException e) {
+      // Do nothing. We've waited up to 1 second for processing to start, if it didn't start in this
       // time the engine probably has not got anything left to process.
     }
   }
@@ -203,5 +152,64 @@ public class Utilities {
       throws InterruptedException, TimeoutException {
     client.newThrowErrorCommand(key).errorCode(errorCode).errorMessage(errorMessage).send().join();
     waitForIdleState(engine, Duration.ofSeconds(1));
+  }
+
+  public static final class ProcessPackLoopingServiceTask {
+
+    public static final String RESOURCE_NAME = "looping-servicetask.bpmn";
+    public static final String PROCESS_ID = "looping-servicetask";
+
+    public static final String ELEMENT_ID = "servicetask"; // id of the service task
+    public static final String JOB_TYPE = "test"; // job type of service task
+    public static final String TOTAL_LOOPS =
+        "totalLoops"; // variable name to indicate number of loops
+    public static final String GATEWAY_ELEMENT_ID = "Gateway_0fhwf5d";
+    public static final String START_EVENT_ID = "startevent";
+    public static final String END_EVENT_ID = "endevent";
+  }
+
+  public static final class ProcessPackMultipleTasks {
+
+    public static final String RESOURCE_NAME = "multiple-tasks.bpmn";
+    public static final String PROCESS_ID = "multiple-tasks";
+    public static final String ELEMENT_ID_1 = "servicetask1";
+    public static final String ELEMENT_ID_2 = "servicetask2";
+    public static final String ELEMENT_ID_3 = "servicetask3";
+  }
+
+  public static final class ProcessPackMessageEvent {
+
+    public static final String RESOURCE_NAME = "message-event.bpmn";
+    public static final String PROCESS_ID = "message-event";
+    public static final String MESSAGE_NAME = "message";
+    public static final String CORRELATION_KEY_VARIABLE = "correlationKey";
+  }
+
+  public static final class ProcessPackMessageStartEvent {
+
+    public static final String RESOURCE_NAME = "message-start-event.bpmn";
+    public static final String MESSAGE_NAME = "start-message";
+    public static final String CORRELATION_KEY = "";
+  }
+
+  public static final class ProcessPackTimerStartEvent {
+
+    public static final String RESOURCE_NAME = "timer-start-event-daily.bpmn";
+    public static final String TIMER_ID = "timer";
+  }
+
+  public static final class ProcessPackCallActivity {
+
+    public static final String RESOURCE_NAME = "call-activity.bpmn";
+    public static final String PROCESS_ID = "call-activity";
+    public static final String CALL_ACTIVITY_ID = "callactivity";
+    public static final String CALLED_RESOURCE_NAME = ProcessPackStartEndEvent.RESOURCE_NAME;
+    public static final String CALLED_PROCESS_ID = ProcessPackStartEndEvent.PROCESS_ID;
+  }
+
+  public static final class ProcessPackStartEndEvent {
+
+    public static final String RESOURCE_NAME = "start-end.bpmn";
+    public static final String PROCESS_ID = "start-end";
   }
 }

